torchmdnet/datasets/custom.py

```diff
@@ -2,6 +2,34 @@
 import numpy as np
 import torch
 from torch_geometric.data import Dataset, Data
+from .hdf import HDF5
+import h5py
+
+__all__ = ["Custom"]
+
+
+def write_as_hdf5(files, hdf5_dataset):
+    """Transform the input to hdf5 format compatible with the HDF5 Dataset class.
+    The input files to the Custom dataset are transformed to a single HDF5 file.
+    Args:
+        files (dict): Dictionary of input files. Must contain "pos", "z" and at least one of "y" or "neg_dy".
+        hdf5_dataset (string): Path to the output HDF5 dataset.
+    """
+    with h5py.File(hdf5_dataset, "w") as f:
+        for i in range(len(files["pos"])):
+            # Create a group for each file
+            coord_data = np.load(files["pos"][i])
+            embed_data = np.load(files["z"][i]).astype(int)
+            group = f.create_group(str(i))
+            num_samples = coord_data.shape[0]
+            group["pos"] = coord_data
+            group["types"] = np.tile(embed_data, (num_samples, 1))
+            if "y" in files:
+                energy_data = np.load(files["y"][i])
+                group["energy"] = energy_data
+            if "neg_dy" in files:
+                force_data = np.load(files["neg_dy"][i])
+                group["forces"] = force_data
```

Comment:
I think the better way to create a hdf5 file (especially for the datasets) is: 
```python 
f = h5py.File(hdf5_dataset, "w")
for i in range(len(files["pos"])):
    # Create a group for each file
    group = f.create_group(str(i))
    coord_data = np.load(files["pos"][i], mmap_mode="r")
    embed_data = np.load(files["z"][i], mmap_mode="r").astype(int)
    num_samples = coord_data.shape[0]
    group.create_dataset("pos", data=coords_data)
    group.create_dataset("types", data=np.tile(embed_data, (num_samples, 1)))
    if "y" in files:
        energy_data = np.load(files["y"][i], mmap_mode="r")
        group.create_dataset("energy", data=energy_data)
    if "neg_dy" in files:
        force_data = np.load(files["neg_dy"][i], mmap_mode="r")
        group.create_dataset("forces", data=force_data)
f.close()
```
I also added `mmap_mode="r"` when you call numpy load. These are just tips, feel free to skip it if you think are not useful.

Reply:
mmap_mode makes sense, as the dataset might be super big.
Why is group.create_dataset("something",... better than group["something"] = ?

Reply:
It's just the "proper way", in theory you should be able to move across path in the hdf5 file (it's not this case) . For example if you use "create_dataset" you can retrieve the pos dataset using:
```python 
f = h5py.File(hdf5_dataset, "r")
f["0/pos"]
```
While if you use the "dictionary way" you get this error: 
`KeyError: "Unable to open object (object 'pos' doesn't exist)"`

Reply:
I have always used that way, assuming was the proper one.  [reference](https://docs.h5py.org/en/stable/high/dataset.html)

Reply:
Convinced! Thanks.
I am not sure repeating the type data is necessary, but I found no way around it:
    group.create_dataset("types", data=np.tile(embed_data, (num_samples, 1)))
Can you think of something?